x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/dataframe/process/ChunkedTrainedModelPersister.java

```diff
@@ -0,0 +1,208 @@
+                        "[{}] finished stored trained model definition chunks with id [{}]",
+                        analytics.getId(),
+                        this.currentModelId.get());
+                    auditor.info(analytics.getId(), "Stored trained model with id [" + this.currentModelId.get() + "]");
+                    CountDownLatch refreshLatch = new CountDownLatch(1);
+                    provider.refreshInferenceIndex(
+                        new LatchedActionListener<>(ActionListener.wrap(
+                            refreshResponse -> LOGGER.debug(() -> new ParameterizedMessage(
+                                "[{}] refreshed inference index after model store",
+                                analytics.getId()
+                            )),
+                            e -> LOGGER.warn("[{}] failed to refresh inference index after model store", analytics.getId())),
+                            refreshLatch));
+                    try {
+                        if (refreshLatch.await(30, TimeUnit.SECONDS) == false) {
+                            LOGGER.error("[{}] Timed out (30s) waiting for index refresh", analytics.getId());
+                        }
+                    } catch (InterruptedException e) {
+                        Thread.currentThread().interrupt();
+                    }
+                }
+            },
+            e -> failureHandler.accept(ExceptionsHelper.serverError("error storing trained model definition chunk [{}] with id [{}]", e,
+                trainedModelDefinitionDoc.getModelId(), trainedModelDefinitionDoc.getDocNum()))
+        );
+        provider.storeTrainedModelDefinitionDoc(trainedModelDefinitionDoc, new LatchedActionListener<>(storeListener, latch));
+        try {
+            if (latch.await(30, TimeUnit.SECONDS) == false) {
+                LOGGER.error("[{}] Timed out (30s) waiting for chunked inference definition to be stored", analytics.getId());
+            }
+        } catch (InterruptedException e) {
+            Thread.currentThread().interrupt();
+            failureHandler.accept(ExceptionsHelper.serverError("interrupted waiting for chunked inference definition to be stored"));
+        }
+    }
+
+    public void createAndIndexInferenceModelMetadata(ModelSizeInfo inferenceModelSize) {
+        if (readyToStoreNewModel == false) {
+            failureHandler.accept(ExceptionsHelper.serverError(
+                "new inference model is attempting to be stored before completion previous model storage"
+            ));
+            return;
+        }
+        TrainedModelConfig trainedModelConfig = createTrainedModelConfig(inferenceModelSize);
+        CountDownLatch latch = storeTrainedModelMetadata(trainedModelConfig);
+        try {
+            readyToStoreNewModel = false;
+            if (latch.await(30, TimeUnit.SECONDS) == false) {
+                LOGGER.error("[{}] Timed out (30s) waiting for inference model metadata to be stored", analytics.getId());
```

Comment:
if this happens, it seems the persister can get stuck, because the doc gets never be stored and `readyToStoreNewModel` is never reset? Correct me if I am wrong.

Reply:
yeah, it should reset. Good catch

Reply:
I do not think it should switch back in this timeout check. If it times out, it just took a long time to persist. 

If the persistence itself fails, then I will reset the boolean flag. 

Similar behavior for the persistence of the definition docs. exception being, if the definition doc is the `eos`, then I will reset the flag.